rust/yarp-sys/tests/serialize_tests.rs

```diff
@@ -0,0 +1,102 @@
+    let ruby_file_path = rust_path.join("../../lib/yarp.rb").canonicalize().unwrap();
+    let file_contents = std::fs::read_to_string(ruby_file_path).unwrap();
+    let len = file_contents.len();
+
+    (CString::new(file_contents).unwrap(), len)
+}
+
+#[test]
+fn serialize_test() {
+    let (ruby_file_contents, len) = ruby_file_contents();
+    let source = ruby_file_contents.as_ptr();
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+    let mut buffer = MaybeUninit::<yp_buffer_t>::uninit();
+
+    unsafe {
+        yp_parser_init(parser.as_mut_ptr(), source, len, std::ptr::null());
+        let parser = parser.assume_init_mut();
+        let node = yp_parse(parser);
+
+        if !yp_buffer_init(buffer.as_mut_ptr()) {
+            panic!("Failed to init buffer");
+        }
+
+        let buffer = buffer.assume_init_mut();
+        yp_serialize(parser, node, buffer);
+
+        let serialized = std::slice::from_raw_parts(buffer.value as *mut u8, buffer.length);
+
+        assert_eq!(&serialized[0..4], b"YARP");
+        assert_eq!(serialized[4..5][0], 0); // YP_VERSION_MAJOR
+        assert_eq!(serialized[5..6][0], 6); // YP_VERSION_MINOR
+        assert_eq!(serialized[6..7][0], 0); // YP_VERSION_PATCH
+
+        yp_buffer_free(buffer);
+        yp_node_destroy(parser, node);
+        yp_parser_free(parser);
+    }
+}
+
+#[test]
+fn parse_serialize_test() {
+    let (ruby_file_contents, len) = ruby_file_contents();
+    let source = ruby_file_contents.as_ptr();
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+    let mut serialize_buffer = MaybeUninit::<yp_buffer_t>::uninit();
+    let mut parse_serialize_buffer = MaybeUninit::<yp_buffer_t>::uninit();
+
+    let serialized = unsafe {
+        yp_parser_init(parser.as_mut_ptr(), source, len, std::ptr::null());
+        let parser = parser.assume_init_mut();
+        let node = yp_parse(parser);
+
+        if !yp_buffer_init(serialize_buffer.as_mut_ptr()) {
+            panic!("Failed to init buffer");
+        }
+
+        let serialize_buffer = serialize_buffer.assume_init_mut();
+        yp_serialize(parser, node, serialize_buffer);
+
+        yp_node_destroy(parser, node);
+        yp_parser_free(parser);
+
+        // Can't use String -> CString here because `value` contains nul bytes.
+        Vec::from_raw_parts(
+            serialize_buffer.value as *mut u8,
+            serialize_buffer.length,
+            serialize_buffer.capacity,
+        )
+    };
+
+    unsafe {
+        if !yp_buffer_init(parse_serialize_buffer.as_mut_ptr()) {
+            panic!("Failed to init buffer");
+        }
+
+        let parse_serialize_buffer = parse_serialize_buffer.assume_init_mut();
+        let metadata = std::ptr::null();
+
+        yp_parse_serialize(
+            serialized.as_ptr() as *const i8,
+            serialized.len(),
+            parse_serialize_buffer,
+            metadata,
+        );
+
+        let string = String::from_raw_parts(
```

Comment:
this takes ownership of the memory in the `parse_serialize_buffer` and will free it when the `String` is dropped. Is that intentional?

Reply:
Nope.